File: console/src/quickstarts/regional_queue/quickstart.py
from __future__ import annotations

import json
import os
from typing import Any

from epoch_sdk import EventEnvelope, RegionalQueueClient, RegionalScope


def result(document: dict[str, Any]) -> dict[str, Any]:
    receipt = document["receipt"]
    assert isinstance(receipt, dict)
    outcome = receipt["outcome"]
    assert isinstance(outcome, dict)
    operation = outcome["result"]
    assert isinstance(operation, dict)
    return operation


client = RegionalQueueClient(
    os.getenv(
        "EPOCH_REGIONAL_ENDPOINTS",
        "http://127.0.0.1:18661,http://127.0.0.1:18662,http://127.0.0.1:18663",
    ).split(","),
    token=os.getenv("EPOCH_TOKEN", "epoch-dev-admin-v1"),
    scope=RegionalScope("acme", "shop", "dev", "core"),
    timeout=3.0,
)
event = EventEnvelope(
    id="docs-python-job-42",
    source="docs-python",
    event_type="job.created",
    payload={"job_id": "python-42"},
    time_ms=42,
)

enqueued = client.enqueue("jobs", 0, "docs-python-enqueue-v1", event)
replayed = client.enqueue("jobs", 0, "docs-python-enqueue-v1", event)
acquired = client.acquire(
    "jobs",
    0,
    "docs-python-acquire-v1",
    consumer="docs-python",
    consumer_epoch=1,
    max_messages=1,
    max_in_flight=1,
    visibility_timeout_ms=5_000,
)
delivery = result(acquired)["deliveries"][0]
extended = client.extend_lease(
    "jobs",
    0,
    "docs-python-extend-v1",
    "docs-python",
    1,
    delivery["lease_token"],
    60_000,
)
released = client.release(
    "jobs",
    0,
    "docs-python-release-v1",
    "docs-python",
    1,
    result(extended)["lease_token"],
    0,
    "demonstrate retry",
)
maintained = client.maintain("jobs", 0, "docs-python-maintain-v1")
reacquired = client.acquire(
    "jobs",
    0,
    "docs-python-reacquire-v1",
    consumer="docs-python",
    consumer_epoch=1,
    max_messages=1,
    max_in_flight=1,
)
redelivery = result(reacquired)["deliveries"][0]
rejected = client.reject(
    "jobs",
    0,
    "docs-python-reject-v1",
    "docs-python",
    1,
    redelivery["lease_token"],
    "poison",
)
history_id = int(result(rejected)["dead_letter_history_id"])
dead_letters = client.dead_letters("jobs", 0, limit=10)
redriven = client.redrive("jobs", 0, "docs-python-redrive-v1", event.id, history_id)
final_acquire = client.acquire(
    "jobs",
    0,
    "docs-python-final-acquire-v1",
    consumer="docs-python",
    consumer_epoch=1,
    max_messages=1,
    max_in_flight=1,
)
final_delivery = result(final_acquire)["deliveries"][0]
acknowledged = client.acknowledge(
    "jobs",
    0,
    "docs-python-ack-v1",
    "docs-python",
    1,
    final_delivery["lease_token"],
)

session_event = EventEnvelope(
    id="docs-python-session-42",
    source="docs-python",
    event_type="session.job.created",
    payload={"job_id": "python-session-42"},
    time_ms=43,
)
session_enqueue = client.enqueue(
    "jobs",
    0,
    "docs-python-session-enqueue-v1",
    session_event,
    session_id="account-python-7",
    correlation_id="request-python-7",
    reply_to="reply-temporary",
)
correlated = client.correlation("jobs", 0, "request-python-7")
session_acquire = client.acquire(
    "jobs",
    0,
    "docs-python-session-acquire-v1",
    consumer="docs-python-session",
    consumer_epoch=1,
    max_messages=1,
    max_in_flight=1,
    visibility_timeout_ms=5_000,
    session_id="account-python-7",
)
session_renew = client.renew_session_lock(
    "jobs",
    0,
    "docs-python-session-renew-v1",
    "docs-python-session",
    1,
    result(session_acquire)["session_lock_token"],
    30_000,
)
client.acknowledge(
    "jobs",
    0,
    "docs-python-session-ack-v1",
    "docs-python-session",
    1,
    result(session_acquire)["deliveries"][0]["lease_token"],
)
session_release = client.release_session_lock(
    "jobs",
    0,
    "docs-python-session-release-v1",
    "docs-python-session",
    1,
    result(session_renew)["session_lock_token"],
)

deferred_event = EventEnvelope(
    id="docs-python-deferred-42",
    source="docs-python",
    event_type="job.deferred",
    payload={"job_id": "python-deferred-42"},
    time_ms=44,
)
client.enqueue("jobs", 0, "docs-python-deferred-enqueue-v1", deferred_event)
deferred_acquire = client.acquire(
    "jobs",
    0,
    "docs-python-deferred-acquire-v1",
    consumer="docs-python-deferred",
    consumer_epoch=1,
    max_messages=1,
    max_in_flight=1,
)
deferred = client.defer(
    "jobs",
    0,
    "docs-python-defer-v1",
    "docs-python-deferred",
    1,
    result(deferred_acquire)["deliveries"][0]["lease_token"],
    "await dependency",
)
received_deferred = client.receive_deferred(
    "jobs",
    0,
    "docs-python-receive-deferred-v1",
    deferred_event.id,
    "docs-python-deferred",
    1,
    visibility_timeout_ms=5_000,
)
client.acknowledge(
    "jobs",
    0,
    "docs-python-deferred-ack-v1",
    "docs-python-deferred",
    1,
    result(received_deferred)["delivery"]["lease_token"],
)

print(
    json.dumps(
        {
            "enqueue": enqueued,
            "exact_retry": replayed,
            "release": released,
            "maintain": maintained,
            "dead_letters": dead_letters,
            "redrive": redriven,
            "ack": acknowledged,
            "counts": client.counts("jobs", 0),
            "flow": client.consumer_flow("jobs", 0, "docs-python"),
            "session_enqueue": session_enqueue,
            "correlation": correlated,
            "session_release": session_release,
            "defer": deferred,
            "receive_deferred": received_deferred,
            "advanced": client.advanced_status("jobs", 0),
            "dead_letter_forwards": client.dead_letter_forwards("jobs", 0, limit=10),
        },
        indent=2,
    )
)

File: console/src/quickstarts/regional_queue/quickstart.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"epoch.local/epoch/sdk/go/epoch"
)

func main() {
	client, err := epoch.NewRegionalQueueClient(
		strings.Split(environment("EPOCH_REGIONAL_ENDPOINTS", "http://127.0.0.1:18661,http://127.0.0.1:18662,http://127.0.0.1:18663"), ","),
		environment("EPOCH_TOKEN", "epoch-dev-admin-v1"),
		epoch.RegionalScope{Organization: "acme", Project: "shop", Environment: "dev", Namespace: "core"},
		3*time.Second,
	)
	must(err)
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()
	event := epoch.NewEventEnvelope("docs-go", "job.created", map[string]any{"job_id": "go-42"})
	event.ID, event.TimeMS = "docs-go-job-42", 42

	enqueued, err := client.Enqueue(ctx, "jobs", 0, "docs-go-enqueue-v1", event)
	must(err)
	replayed, err := client.Enqueue(ctx, "jobs", 0, "docs-go-enqueue-v1", event)
	must(err)
	window, timeout := uint16(1), uint64(5_000)
	acquired, err := client.Acquire(ctx, "jobs", 0, "docs-go-acquire-v1", epoch.RegionalQueueAcquireOptions{
		Consumer: "docs-go", ConsumerEpoch: 1, MaxMessages: 1,
		MaxInFlight: &window, VisibilityTimeoutMS: &timeout,
	})
	must(err)
	delivery := result(acquired)["deliveries"].([]any)[0].(map[string]any)
	extended, err := client.ExtendLease(ctx, "jobs", 0, "docs-go-extend-v1", "docs-go", 1, delivery["lease_token"].(string), 60_000)
	must(err)
	released, err := client.Release(ctx, "jobs", 0, "docs-go-release-v1", "docs-go", 1, result(extended)["lease_token"].(string), 0, "demonstrate retry")
	must(err)
	maintained, err := client.Maintain(ctx, "jobs", 0, "docs-go-maintain-v1")
	must(err)
	reacquired, err := client.Acquire(ctx, "jobs", 0, "docs-go-reacquire-v1", epoch.RegionalQueueAcquireOptions{Consumer: "docs-go", ConsumerEpoch: 1, MaxMessages: 1, MaxInFlight: &window})
	must(err)
	redelivery := result(reacquired)["deliveries"].([]any)[0].(map[string]any)
	rejected, err := client.Reject(ctx, "jobs", 0, "docs-go-reject-v1", "docs-go", 1, redelivery["lease_token"].(string), "poison")
	must(err)
	historyID, err := strconv.ParseUint(fmt.Sprint(result(rejected)["dead_letter_history_id"]), 10, 64)
	must(err)
	deadLetters, err := client.DeadLetters(ctx, "jobs", 0, 10)
	must(err)
	redriven, err := client.Redrive(ctx, "jobs", 0, "docs-go-redrive-v1", event.ID, historyID)
	must(err)
	finalAcquire, err := client.Acquire(ctx, "jobs", 0, "docs-go-final-acquire-v1", epoch.RegionalQueueAcquireOptions{Consumer: "docs-go", ConsumerEpoch: 1, MaxMessages: 1, MaxInFlight: &window})
	must(err)
	finalDelivery := result(finalAcquire)["deliveries"].([]any)[0].(map[string]any)
	acknowledged, err := client.Acknowledge(ctx, "jobs", 0, "docs-go-ack-v1", "docs-go", 1, finalDelivery["lease_token"].(string))
	must(err)

	sessionID, correlationID, replyTo := "account-go-7", "request-go-7", "reply-temporary"
	sessionEvent := epoch.NewEventEnvelope("docs-go", "session.job.created", map[string]any{"job_id": "go-session-42"})
	sessionEvent.ID, sessionEvent.TimeMS = "docs-go-session-42", 43
	sessionEnqueue, err := client.EnqueueAdvanced(ctx, "jobs", 0, "docs-go-session-enqueue-v1", sessionEvent, epoch.RegionalQueueEnqueueOptions{
		SessionID: sessionID, CorrelationID: correlationID, ReplyTo: replyTo,
	})
	must(err)
	correlated, err := client.Correlation(ctx, "jobs", 0, correlationID)
	must(err)
	sessionAcquire, err := client.Acquire(ctx, "jobs", 0, "docs-go-session-acquire-v1", epoch.RegionalQueueAcquireOptions{
		Consumer: "docs-go-session", ConsumerEpoch: 1, MaxMessages: 1,
		MaxInFlight: &window, VisibilityTimeoutMS: &timeout, SessionID: sessionID,
	})
	must(err)
	sessionResult := result(sessionAcquire)
	sessionRenew, err := client.RenewSessionLock(ctx, "jobs", 0, "docs-go-session-renew-v1", "docs-go-session", 1, sessionResult["session_lock_token"].(string), 30_000)
	must(err)
	sessionDelivery := sessionResult["deliveries"].([]any)[0].(map[string]any)
	_, err = client.Acknowledge(ctx, "jobs", 0, "docs-go-session-ack-v1", "docs-go-session", 1, sessionDelivery["lease_token"].(string))
	must(err)
	sessionRelease, err := client.ReleaseSessionLock(ctx, "jobs", 0, "docs-go-session-release-v1", "docs-go-session", 1, result(sessionRenew)["session_lock_token"].(string))
	must(err)

	deferredEvent := epoch.NewEventEnvelope("docs-go", "job.deferred", map[string]any{"job_id": "go-deferred-42"})
	deferredEvent.ID, deferredEvent.TimeMS = "docs-go-deferred-42", 44
	_, err = client.Enqueue(ctx, "jobs", 0, "docs-go-deferred-enqueue-v1", deferredEvent)
	must(err)
	deferredAcquire, err := client.Acquire(ctx, "jobs", 0, "docs-go-deferred-acquire-v1", epoch.RegionalQueueAcquireOptions{Consumer: "docs-go-deferred", ConsumerEpoch: 1, MaxMessages: 1, MaxInFlight: &window})
	must(err)
	deferredDelivery := result(deferredAcquire)["deliveries"].([]any)[0].(map[string]any)
	deferred, err := client.Defer(ctx, "jobs", 0, "docs-go-defer-v1", "docs-go-deferred", 1, deferredDelivery["lease_token"].(string), "await dependency")
	must(err)
	receivedDeferred, err := client.ReceiveDeferred(ctx, "jobs", 0, "docs-go-receive-deferred-v1", deferredEvent.ID, "docs-go-deferred", 1, &timeout)
	must(err)
	_, err = client.Acknowledge(ctx, "jobs", 0, "docs-go-deferred-ack-v1", "docs-go-deferred", 1, result(receivedDeferred)["delivery"].(map[string]any)["lease_token"].(string))
	must(err)
	advanced, err := client.AdvancedStatus(ctx, "jobs", 0)
	must(err)
	forwards, err := client.DeadLetterForwards(ctx, "jobs", 0, 10)
	must(err)
	counts, err := client.Counts(ctx, "jobs", 0)
	must(err)
	flow, err := client.ConsumerFlow(ctx, "jobs", 0, "docs-go")
	must(err)

	output, err := json.MarshalIndent(map[string]any{
		"enqueue": enqueued, "exact_retry": replayed, "release": released, "maintain": maintained,
		"dead_letters": deadLetters, "redrive": redriven, "ack": acknowledged, "counts": counts, "flow": flow,
		"session_enqueue": sessionEnqueue, "correlation": correlated, "session_release": sessionRelease,
		"defer": deferred, "receive_deferred": receivedDeferred, "advanced": advanced, "dead_letter_forwards": forwards,
	}, "", "  ")
	must(err)
	fmt.Println(string(output))
}

func result(document epoch.Document) map[string]any {
	receipt := document["receipt"].(map[string]any)
	outcome := receipt["outcome"].(map[string]any)
	return outcome["result"].(map[string]any)
}

func environment(name, fallback string) string {
	if value := strings.TrimSpace(os.Getenv(name)); value != "" {
		return value
	}
	return fallback
}

func must(err error) {
	if err != nil {
		panic(err)
	}
}
